eth/stagedsync/stage_bodies.go

```diff
@@ -119,137 +119,56 @@ func BodiesForward(
-				start = time.Now()
-				cfg.bd.RequestSent(req, currentTime+uint64(timeout), peer)
-				d3 += time.Since(start)
-				log.Debug("body request sent", "req", fmt.Sprintf("%+v", req), "peer", fmt.Sprintf("%x", peer))
-			}
 
-			loopCount++
-			if loopCount >= requestLoopCutOff {
-				break
-			}
+		d1, d2, d3, err = requestMoreBodies(ctx, d1, d2, d3, start, cfg, tx, timeout)
+		if err != nil {
+			return false, err
 		}
 
 		start = time.Now()
-		requestedLow, delivered, err := cfg.bd.GetDeliveries(tx)
+		var requestedLow, delivered uint64
+		requestedLow, delivered, err = cfg.bd.GetDeliveries(tx)
 		if err != nil {
 			return false, err
 		}
 		totalDelivered += delivered
 		d4 += time.Since(start)
+		td4 = time.Since(start)
+		if delivered > 0 {
+			log.Debug("Timings GetDeliveries", "delivered", delivered, "td4", td4)
+		}
 		start = time.Now()
 		cr := ChainReader{Cfg: cfg.chanConfig, Db: tx}
-
-		toProcess := cfg.bd.NextProcessingCount()
-
-		if toProcess > 0 {
-			var i uint64
-			for i = 0; i < toProcess; i++ {
-				nextBlock := requestedLow + i
-
-				header, _, err := cfg.bd.GetHeader(nextBlock, cfg.blockReader, tx)
-				if err != nil {
-					return false, err
-				}
-				blockHeight := header.Number.Uint64()
-				if blockHeight != nextBlock {
-					return false, fmt.Errorf("[%s] Header block unexpected when matching body, got %v, expected %v", logPrefix, blockHeight, nextBlock)
-				}
-
-				rawBody := cfg.bd.GetBodyFromCache(nextBlock)
-				if rawBody == nil {
-					log.Debug("Body was nil when reading from cache", "block", nextBlock)
-					break
-				}
-
-				// Txn & uncle roots are verified via bd.requestedMap
-				err = cfg.bd.Engine.VerifyUncles(cr, header, rawBody.Uncles)
-				if err != nil {
-					log.Error(fmt.Sprintf("[%s] Uncle verification failed", logPrefix), "number", blockHeight, "hash", header.Hash().String(), "err", err)
-					u.UnwindTo(blockHeight-1, header.Hash())
-					return true, nil
-				}
-
-				// Check existence before write - because WriteRawBody isn't idempotent (it allocates new sequence range for transactions on every call)
-				ok, lastTxnNum, err := rawdb.WriteRawBodyIfNotExists(tx, header.Hash(), blockHeight, rawBody)
-				if err != nil {
-					return false, fmt.Errorf("WriteRawBodyIfNotExists: %w", err)
-				}
-				if cfg.historyV3 && ok {
-					if err := rawdb.TxNums.Append(tx, blockHeight, lastTxnNum); err != nil {
-						return false, err
-					}
-				}
-
-				if blockHeight > bodyProgress {
-					bodyProgress = blockHeight
-					if err = s.Update(tx, blockHeight); err != nil {
-						return false, fmt.Errorf("saving Bodies progress: %w", err)
-					}
-				}
-				cfg.bd.AdvanceLow()
+		var processedCount uint64 = 0
+		for {
+			toProcess := cfg.bd.NextProcessingCount()
+			if toProcess == 0 {
+				break
+			}
+			var skip bool
+			bodyProgress, skip, err = processNextBodies(cr, toProcess, requestedLow, cfg, tx, logPrefix, bodyProgress, s, u)
+			if skip || err != nil {
+				return skip, err
```

Comment:
if `skip == true` but `err == nil`, this will not hit `tx.Commit()` later, but it should

Reply:
Instead there should be separate `if err != nil` and then `if skip`

Reply:
I think this would be OK?  Skip just breaks the infinite for loop that `loopBody` is called and a commit will happen after that if we're not in an external transaction.  I've updated the code regardless to `return false, err` in the case of an error.  Looking at that code I've noticed i missed an error check a few lines below so I'm sorting that too.